### DISCUSSION TITLE
A proper AML conformance test suite

### DISCUSSION THREAD

**Discussion by d-tatianin** [acpica/acpica#956] [Ideas]
Hi there!

I'm very interesting in developing a shared test suite that can be used to ensure AML/ACPI specification conformance of a given interpreter, very similar to the javascript's test-262 suite.

Currently some of ACPICA tests use a special AML<->runner ABI for propagating exceptions to AML and are therefore very coupled with ACPICA. I was thinking maybe we could come up with an implementation abstraction layer that defines a set of predefined helpers that a test AML blob can use/call into to communicate with the test runner so that the tests can be ported between implementations?

Is there any interest in moving the test suite outside of the ACPICA's repository and making it more of a generic test suite?
I think this can be beneficial for improving the coverage as well as ensuring an implementation is correct in terms of what AML expects, especially since the specification is either extremely vague, or just straight up doesn't cover a lot of edge cases or e.g. reference semantics whatsoever

My [ACPI subsystem](https://github.com/UltraOS/uACPI) has tons of its own [tests](https://github.com/UltraOS/uACPI/tree/master/tests), some of which can definitely be beneficial for ACPICA as well. For example, it has an automated way to run about 500 [real hardware AML blobs](https://github.com/UltraOS/ACPIDumps) against the runner (which I do on every commit), not to mention other functionality-specific tests.

In my opinion if we want to bring ACPI/AML to a point where it's more standardized and uniform across implementations -- this is the way forward.

Thoughts?

Thank you!

**Comment by SaketADumbre** [comment]
Thank you very much @d-tatianin for opening this conversation, the idea sounds very interesting! We can figure out the finer details in the near future, but I am definitely interested in this idea. The ASL Test Suite is fairly old at this point and was never finished completely due to the Russian team working on it at Intel back then (maybe 15-20 years ago or so) got shelved or reassigned to other work midway through their work on the ASLTS. I can revisit this thread in about a couple more weeks when I am fully settled down again in my job after a long time off, but please keep the conversation going and talking about the initial interest and possibly the overall architecture ideas. Thanks a lot!
